kale/embed/mpca.py

```diff
@@ -1,181 +1,308 @@
+"""
+
+import sys
 import numpy as np
-import tensorly as tl
-from tensorly.base import fold, unfold
-from tensorly.tenalg import mode_dot, multi_mode_dot
+# import tensorly as tl
+from tensorly.base import unfold
+from tensorly.tenalg import multi_mode_dot
 from sklearn.base import BaseEstimator, TransformerMixin
 
 
+def check_ndim(X, ndim):
+    """Check if the order/mode of the input data is consistent with the given number of order/mode
+
+    Args:
+        X (ndarray): input data
+        ndim (int): number of order/mode expected
+
+    Returns:
+        boolean value: True if the order of X is consistent with ndim, False vise versa
+    """
+
+    return X.ndim == ndim
+
+
+def check_shape(X, shape_):
+    """Check if the shape (excluding the last mode) of the input data is consistent with the given shape
+
+    Args:
+        X (ndarray): input data
+        shape_: shape expected
+
+    Returns:
+        boolean value: True if the shape of X is consistent with shape_, False vise versa
+    """
+
+    return X.shape[:-1] == shape_
+
+
+def _check_dim_shape(X, ndim, shape_):
+    """Check if the order/mode of the input data is consistent with the given number of order/mode
+
+    Args:
+        X (ndarray): input data
+        ndim (int): number of order/mode expected
+        shape_: shape expected
+
+    """
+    if not check_ndim(X, ndim) or not check_shape(X, shape_):
+        print("The given data should be consistent with the order and shape of training data")
+        sys.exit()
+
+
 class MPCA(BaseEstimator, TransformerMixin):
-    def __init__(self, variance_explained=0.97, max_iter=1):
-        """
-        Implementation compatible with sickit-learn
-        ----------
-        Parameter:
-            variance_explained: ration of variance to keep (between 0 and 1)
-            max_iter: max number of iteration, integer
-        ----------
-        Attributes:
-            tPs: list of projection matrices
-        """
-        self.var_exp = variance_explained
-        self.max_iter = max_iter
-        # self.Xmean = None
-        # self.n_dim = None
-        self.tPs = []
+    """MPCA implementation compatible with sickit-learn
 
-    def fit(self, X):
-        """
-        Parameter:
-            X: array-like, ndarray of shape (I1, I2, ..., n_samples)
-        ----------
-        Return:
-            self
+    Args:
+        var_explained (float, optional): amount of variance to keep
+            (between 0 and 1). Defaults to 0.97.
+        max_iter (int, optional): max number of iteration. Defaults to 1.
+
+    Attributes:
+        proj_mats (list): a list of transposed projection matrices
+        idx_order (ndarray): the ordering index of projected (and vectorised) features in decreasing variance
+        shape_in (tuple): input tensor shapes, i.e. (P_1, P_2, ..., P_N)
+        shape_out (tuple): output tensor shapes, i.e. (i_1, i_2, ..., i_N)
```

Comment:
I am confused here. Typically, we use I_1, ..., I_N to denote the input dimensions and P_1, ..., P_N to denote the output dimensions. Why using opposite notations here?
Also i-n should be capitalised to I_n.

Reply:
Fixed. Please see the latest update.